### PR TITLE
Entity Web API generation

### PR DESCRIPTION
**Pull request by bohdan-harniuk**
**Description** (*)
This PR adds the possibility of generating REST API endpoints for managing entity generated with the entity creator feature.

**What was done:**
1. added to entity creator dialog checkbox that indicates if REST API endpoints should be created (by default is disabled)
2. generates API for the DeleteByIdCommand class, adds declaration to the webapi.xml file, replaces the original class type on the newly created interface whenever it is used
3. generates API for the SaveCommand class, adds declaration to the webapi.xml file, replaces the original class type on the newly created interface whenever it is used
4. generates API for the GetListQuery class, adds declaration to the webapi.xml file, replaces the original class type on the newly created interface whenever it is used
5. edited tests that related to the affected parts of the functionality

Screenshot 2021-06-07 at 09 08 35
Screenshot 2021-06-07 at 09 18 54
Screenshot 2021-06-07 at 09 18 00
Screenshot 2021-06-07 at 09 17 25
Screenshot 2021-06-07 at 09 15 49

**Contribution checklist** (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with integration/functional tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: src/com/magento/idea/magento2plugin/actions/generation/data/DeleteEntityControllerFileData.java [hunk 1]
```diff
@@ -26,12 +27,14 @@ public DeleteEntityControllerFileData(
             final @NotNull String entityName,
             final @NotNull String moduleName,
             final @NotNull String acl,
-            final @NotNull String entityId
+            final @NotNull String entityId,
+            final boolean isDeleteCommandHasInterface
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/actions/generation/data/DeleteEntityControllerFileData.java:31]: Let's replace this one to a more readable version: `hasDeleteCommandInterface` or `deleteCommandHasInterface`.

```suggestion
            final boolean hasDeleteCommandInterface
```

File: src/com/magento/idea/magento2plugin/actions/generation/data/SaveEntityControllerFileData.java [hunk 2]
```diff
@@ -35,7 +37,8 @@ public SaveEntityControllerFileData(
             final @NotNull String entityId,
             final @NotNull String dtoName,
             final @NotNull String dtoInterfaceName,
-            final boolean isDtoWithInterface
+            final boolean isDtoWithInterface,
+            final boolean isSaveCommandHasInterface
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/actions/generation/data/SaveEntityControllerFileData.java:41]: Same here.

File: src/com/magento/idea/magento2plugin/actions/generation/data/UiComponentDataProviderData.java [hunk 3]
```diff
@@ -32,17 +33,20 @@ public UiComponentDataProviderData(
...
      * @param entityIdFieldName String
+     * @param isQueryHasInterface boolean
      */
     public UiComponentDataProviderData(
             final String name,
             final String path,
             final String entityName,
-            final String entityIdFieldName
+            final String entityIdFieldName,
+            final boolean isQueryHasInterface
```
Comment (eduard13) [inline src/com/magento/idea/magento2plugin/actions/generation/data/UiComponentDataProviderData.java:43]: Same here

**Review by eduard13 — CHANGES_REQUESTED** [review]
Hi @bohdan-harniuk, thank you for this cool improvement. Please check the below suggestions, and additionally I have a concern regarding the checkbox and what happens if you don't select the pointed checkboxes.

image

So, basically we should have them as dependency or make them individually, because the end-user may not need the Admin UiComponents, but needs only the WebAPI required files.

Please let me know if there are any questions.
Thank you.

File: src/com/magento/idea/magento2plugin/actions/generation/data/DeleteEntityControllerFileData.java (hunk 1 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/actions/generation/data/DeleteEntityControllerFileData.java:31]: Done.

File: src/com/magento/idea/magento2plugin/actions/generation/data/UiComponentDataProviderData.java (hunk 3 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/actions/generation/data/UiComponentDataProviderData.java:43]: Done.

**Comment by bohdan-harniuk** [comment]
Hello, @eduard13! Thank you for the great ideas in your code review!

**What was done:**

1. fixed bug when creating entity without DTO interface (generating was stuck on the data mapper generator)
2. changed bool parameters names (I had to name methods as IsHas... because of AvoidFieldNameMatchingMethodName PMD rule)
3. added commands, query and data mapper generation when creating admin UI components checkbox is unchecked

Thank you!!! Now it looks better!

Could you please proceed with the code review?

**Review by eduard13 — APPROVED** [review]
Good job 👍

**Pull request merged**